x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportOpenJobAction.java

```diff
@@ -700,6 +704,18 @@ void setMaxMachineMemoryPercent(int maxMachineMemoryPercent) {
         }
     }
 
+    // Visible for testing
+    static void checkJobWithRulesRequiresMinVersionOnAllNodes(String jobId, ClusterState clusterState) {
```

Comment:
@droberts195 Actually, I think this is not enough. In the case of a job relocating, `validate` will not be called, right? So if at that point there's a node running on a version earlier than 6.4 it may be assigned to it. I think I need to add a check in the node selection code as well.

Reply:
Yes, if we go down the road of banning use of functionality until the entire cluster is on a particular version then we need to do it when the job is created, not when it's opened.

If we can't do this for some reason and need to do a check when opening a job then it needs to be in the allocation decision, so we might as well then use our existing functionality that takes job version into account when allocating to nodes.

Reply:
The problem with checking when the job is created is we then also need to ban earlier nodes from joining the cluster. It feels too excessive to do that for this. I'll switch to rejecting incompatible nodes from taking on jobs with rules.